### ISSUE TITLE
Compatibility between Other Models and Embedding Models

### ISSUE THREAD

**Issue by Carol-gutianle** [mignonjia/TS_watermark#5]
Dear Authors,
Thank you for your inspiring work!
However, I encountered some issues when I try to use TS in my custom LLM and embedding model.
![image](https://github.com/user-attachments/assets/268da7e5-e3f6-4fae-9e52-5409ac577032)
When using trained gamma and delta network, we can transform input_ids tokenized by other tokenizer to the input_ids tokenized by opt_tokenizer. But if I want to train my custom gamma and network, how can I process these two lines? Because the vocabulary size of my model is always smaller than that of embedding model. It may cause some issues if I try to truncate it to the vocabulary size of embedding model. I am looking forward to your help.
Best,
Carol.

**Comment by mignonjia** [comment]
This training can work because the OPT model and SimCSE model have the same tokenizer. If the tokenizers are different, it might be hard to perform the training, since you need to do a one-to-one mapping between two tokenizers, which is unlikely to be possible.

I would suggest you to try to use PPL of the watermarked text as the loss function of semantics, instead of sentence similarity between non-watermarked and watermarked text. If using PPL, no extra model is needed, and you can just use the original logits and modified logits to compute PPL. (It's just an idea, not sure whether it would work or not)

**Comment by Carol-gutianle** [comment]
Thank you for your prompt response!
But I have one more question, why do I need to use mix_one_hot here to concate the next token embedding to the previous embedding? Why can't we just use the hard_one_hot?
![image](https://github.com/user-attachments/assets/896fb63f-5a39-4b9a-878f-57c5ab871993)
Thank you for your help again!

**Comment by mignonjia** [comment]
soft_one_hot is needed to maintain gradients for gamma/delta network
hard_one_hot doesn't contain gradients, it is to enhance the embedding with the most possible next-token.
So I just mixed them.

**Comment by Carol-gutianle** [comment]
Thank you, I fully understand.
